onnxruntime/core/platform/windows/device_discovery.cc

```diff
@@ -362,90 +362,111 @@ std::unordered_map<uint64_t, DeviceInfo> GetDeviceInfoD3D12() {
   return device_info;
 }
 
+typedef HRESULT(WINAPI* PFN_DXCoreCreateAdapterFactory)(REFIID riid, void** ppvFactory);
+
 // returns LUID to DeviceInfo
 std::unordered_map<uint64_t, DeviceInfo> GetDeviceInfoDxcore() {
   std::unordered_map<uint64_t, DeviceInfo> device_info;
 
-  // Get all GPUs and NPUs by querying WDDM/MCDM.
-  wil::com_ptr<IDXCoreAdapterFactory> adapterFactory;
-  if (FAILED(DXCoreCreateAdapterFactory(IID_PPV_ARGS(&adapterFactory)))) {
+  // Load dxcore.dll. We do this manually so there's not a hard dependency on dxcore which is newer.
+  HMODULE dxcore_lib = LoadLibraryExW(L"dxcore.dll", nullptr, LOAD_LIBRARY_SEARCH_SYSTEM32);
```

Comment:
nit: it would be safer to use an RAII type like unique_ptr to handle the FreeLibrary call. it's not obvious to me that the rest of the code before the existing FreeLibrary call will not throw or return early.

Reply:
WIL has such helpers